util/s3_factory.py

```diff
@@ -0,0 +1,120 @@
+        self._region = _region
+        self._credentials = _credentials or {}
+
+    def download(self, s3_bucket, document_s3_path, version_id=None):
+        """
+        Download object from S3.
+
+        :param s3_bucket: bucket
+        :param document_s3_path: s3 path
+        :param version_id:
+        :return:
+        """
+        try:
+            s3 = boto3.resource("s3", region_name=self._region, **self._credentials)
+            if version_id:
+                instances_file_content = s3.ObjectVersion(s3_bucket, document_s3_path, version_id).get()["Body"].read()
+            else:
+                instances_file_content = s3.Object(s3_bucket, document_s3_path).get()["Body"].read()
+            return instances_file_content
+        except Exception as e:
+            error(
+                "Failed when downloading file %s from bucket %s in region %s with error %s".format(
+                    document_s3_path, s3_bucket, self._region, e
+                )
+            )
+            raise
+
+    def upload(self, s3_bucket, s3_key, data, dryrun=True):
+        """
+        Upload a document to S3.
+
+        :param s3_bucket: bucket
+        :param s3_key: s3 key
+        :param data: byte encoded data
+        :param dryrun: don't actually upload, just print and exit
+        """
+        try:
+            if not dryrun:
+                object = boto3.resource("s3", region_name=self._region, **self._credentials).Object(s3_bucket, s3_key)
+                object.put(Body=data, ACL="public-read")
+            else:
+                logging.info(
+                    "Dryrun mode enabled. The following file would have been uploaded to s3://%s/%s:\n%s",
+                    s3_bucket,
+                    s3_key,
+                    data,
+                )
+        except Exception as e:
+            error(
+                "Failed when uploading file %s to bucket %s in region %s with error %s".format(
+                    s3_key, s3_bucket, self._region, e,
+                )
+            )
+            raise
+
+    def version_exists(self, s3_bucket, document_s3_path):
+        """
+        Sees if the object exists.
+
+        :param s3_bucket: bucket
+        :param document_s3_path: key_path
+        :return: True if exists else False
+        """
+        try:
+            self.get_current_version(s3_bucket, document_s3_path)
+            return True
+        except ClientError:
+            return False
+
+    def get_current_version(self, s3_bucket, document_s3_path):
+        """
+        Get the most recent version of the s3 file.
+
+        :param s3_bucket: bucket
+        :param document_s3_path: s3 key path to the object
+        :return: versionId if it's versioned, else None
+        """
+        response = boto3.client("s3", region_name=self._region, **self._credentials).head_object(
+            Bucket=s3_bucket, Key=document_s3_path
+        )
+        if not response.get("VersionId"):
```

Comment:
I think the default behavior should  be to raise an exception here.

Reply:
changed